core/state/state_override.go

```diff
@@ -0,0 +1,266 @@
+func (db *OverrideDatabase) Reader(root common.Hash) (Reader, error) {
+	reader, err := db.Database.Reader(root)
+	if err != nil {
+		return nil, err
+	}
+	return newOverrideReader(db.overrides, reader), nil
+}
+
+// ContractCode retrieves a particular contract's code.
+func (db *OverrideDatabase) ContractCode(addr common.Address, codeHash common.Hash) ([]byte, error) {
+	override, ok := db.overrides[addr]
+	if ok && override.Code != nil {
+		return *override.Code, nil
+	}
+	return db.Database.ContractCode(addr, codeHash)
+}
+
+// ContractCodeSize retrieves a particular contracts code's size.
+func (db *OverrideDatabase) ContractCodeSize(addr common.Address, codeHash common.Hash) (int, error) {
+	override, ok := db.overrides[addr]
+	if ok && override.Code != nil {
+		return len(*override.Code), nil
+	}
+	return db.Database.ContractCodeSize(addr, codeHash)
+}
+
+// The stateWrapReader wraps a live state database as the state source.
+type stateWrapReader struct {
+	state *StateDB
+}
+
+// Account implementing Reader interface, retrieving the account associated with
+// a particular address.
+//
+// - A nil account is returned if it's not existent
+// - An aggregated error will be returned if readers are all failed
+// - The returned account is safe to modify
+func (r *stateWrapReader) Account(addr common.Address) (*types.StateAccount, error) {
+	obj := r.state.getStateObject(addr)
+	if obj == nil {
+		return nil, nil
+	}
+	return obj.data.Copy(), nil
+}
+
+// Storage implementing Reader interface, retrieving the storage slot associated
+// with a particular account address and slot key.
+//
+// - An empty slot is returned if it's not existent
+// - An aggregated error will be returned if readers are failed
+// - The returned storage slot is safe to modify
+func (r *stateWrapReader) Storage(addr common.Address, storageRoot common.Hash, slot common.Hash) (common.Hash, error) {
+	return r.state.GetState(addr, slot), nil
+}
+
+// Copy implementing Reader interface, returning a deep-copied state reader.
+func (r *stateWrapReader) Copy() Reader {
+	return &stateWrapReader{state: r.state.Copy()}
+}
+
+// stateWrap is an internal struct wrapping a live state instance as the state
+// data source. This can be useful in scenarios where we need to create the
+// snapshot of the current state and apply some additional overrides on top
+// (e.g. state override in RPC call serving).
+type stateWrap struct {
+	Database
+	state *StateDB
+}
+
+// Reader returns a state reader associated with the specified state root.
+func (db *stateWrap) Reader(root common.Hash) (Reader, error) {
+	if root != db.state.originalRoot {
+		return nil, errors.New("state root is not matched")
+	}
+	return &stateWrapReader{state: db.state}, nil
+}
+
+// OverrideState applies the state overrides into the provided live state database.
+func OverrideState(state *StateDB, overrides map[common.Address]OverrideAccount) (*StateDB, error) {
```

Comment:
Can we keep the StateOverride type alias? more cumbersome to type map[common.Address]OverrideAccount 😄